src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -767,12 +804,45 @@ void enforceNotNullWhenForPendingReturn(PendingBranch pendingReturn, BoundReturn
                 {
                     if (pendingReturn.IsConditionalState)
                     {
-                        enforceParameterNotNullWhen(returnStatement.Syntax, parameters, sense: true, stateWhen: pendingReturn.StateWhenTrue);
-                        enforceParameterNotNullWhen(returnStatement.Syntax, parameters, sense: false, stateWhen: pendingReturn.StateWhenFalse);
+                        if (returnStatement.ExpressionOpt is { ConstantValue: { IsBoolean: true, BooleanValue: bool value } })
+                        {
+                            enforceParameterNotNullWhen(returnStatement.Syntax, parameters, sense: value, stateWhen: pendingReturn.State);
+                            return;
+                        }
+
+                        if (!pendingReturn.StateWhenTrue.Reachable || !pendingReturn.StateWhenFalse.Reachable)
```

Comment:
Shouldn't we only suppress when _both_ states are unreachable? We could have just done a test and thrown in the other branch, but accidentally messed up the order of the branches and returned incorrectly.

Reply:
Same for the other tests of this condition.

---
In reply to: [503602424](https://github.com/dotnet/roslyn/pull/48425#discussion_r503602424) [](ancestors = 503602424)

Reply:
The only way to have one state being unreachable is using a constant condition (which is case above).

---
In reply to: [503602558](https://github.com/dotnet/roslyn/pull/48425#discussion_r503602558) [](ancestors = 503602558,503602424)